zkevm-circuits/src/state_circuit/constraint_builder.rs

```diff
@@ -0,0 +1,326 @@
+                "previous tag is Start or address change is 0 or 1",
+                (q.prev_tag.clone() - RwTableTag::Start.expr())
+                    * q.address_change()
+                    * (1.expr() - q.address_change()),
+            )
+        });
+    }
+
+    fn build_account_storage_constraints(&mut self, q: &Queries<F>) {
+        // TODO: cold VS warm
+        // TODO: connection to MPT on first and last access for each (address, key)
+        // No longer true because we moved id from aux to here.
+        // self.require_zero("id is 0 for AccountStorage", q.id());
+        self.require_zero("field_tag is 0 for AccountStorage", q.field_tag());
+        // for every first access, we add an AccountStorage write to setup the
+        // value from the previous block with rw_counter = 0
+        // needs some work...
+        // self.condition(q.first_access(), |cb| {
+        //     cb.require_zero("first access is a write", q.is_write());
+        //     // cb.require_zero("first access rw_counter is 0",
+        // q.rw_counter.value.clone()); })
+    }
+    fn build_tx_access_list_account_constraints(&mut self, q: &Queries<F>) {
+        self.require_zero("field_tag is 0 for TxAccessListAccount", q.field_tag());
+        self.require_zero(
+            "storage_key is 0 for TxAccessListAccount",
+            q.storage_key.encoded.clone(),
+        );
+        // TODO: Missing constraints
+    }
+
+    fn build_tx_access_list_account_storage_constraints(&mut self, q: &Queries<F>) {
+        self.require_zero(
+            "field_tag is 0 for TxAccessListAccountStorage",
+            q.field_tag(),
+        );
+        // TODO: Missing constraints
+    }
+
+    fn build_tx_refund_constraints(&mut self, q: &Queries<F>) {
+        self.require_zero("address is 0 for TxRefund", q.address.value.clone());
+        self.require_zero("field_tag is 0 for TxRefund", q.field_tag());
+        self.require_zero(
+            "storage_key is 0 for TxRefund",
+            q.storage_key.encoded.clone(),
+        );
+        // TODO: Missing constraints
+    }
+
+    fn build_account_constraints(&mut self, q: &Queries<F>) {
+        self.require_zero("id is 0 for Account", q.id());
+        self.require_zero(
+            "storage_key is 0 for Account",
+            q.storage_key.encoded.clone(),
+        );
+        self.require_in_set(
+            "field_tag in AccountFieldTag range",
+            q.field_tag(),
+            set::<F, AccountFieldTag>(),
+        );
+        // // for every first access, we add an Account write to setup the value
+        // from the // previous block with rw_counter = 0
+        // self.condition(q.first_access(), |cb| {
+        //     // cb.require_zero("first access is a write", q.is_write());
+        //     cb.require_zero("first access rw_counter is 0",
+        // q.rw_counter.value.clone()); });
+    }
+
+    fn build_account_destructed_constraints(&mut self, q: &Queries<F>) {
```

Comment:
A similar method for when an account is created is probably needed?
Also, an MPT lookup is needed here too, right (whenever something changes in the state trie)?

Reply:
These are the constraints for rows that match `Rw::AccountDestructed`. I believe that up to three `Rw::Account` will exist in the rw table for account creations? The constraints for these rows are in the `build_account_constraints` function.

I agree MPT lookup(s) will be required here. I need to review the specs to see what our plans (if any) are for handling account desctruction rows.

Reply:
Yes, for the account creation I think we would have three modular changes - changing only one field at a time: nonce, balance, codehash. And we would need to add constraints for the order of these three changes. Is that right @han0110 @ed255 ?

But I think we could leave this for the future PR.

Reply:
Regarding the account destructed, I remember from a discussion with @han0110 that he mentioned that after a SELFDESTRUCT the account is not immediately deleted.  Looking at the yellowpaper I see:

> `SELFDESTRUCT`: Halt execution and register account for later deletion.

So an account destructed row is not directly connected to the MPT, but we need to figure out a way to keep track of all pending account deletes, and at the step EndTx process all those deletions.  Maybe we need to introduce a new tag in the state that signals an account deletion (which must match a previous non-reverted account destruction).  Currently that's open to be specified.

Regarding account creation, I guess there are several ways to create an account:
1. Send balance to an empty account.
2. Tx with `to = 0x0`, CREATE, CREATE2

1 and 2 set different account fields.
I would assume that a non-existing account has the fields to a default value, and on account creation some of those fields are set to a different value.  This makes me wonder if we need to explicitly set all the fields, or just the ones that get a non-default value; and I also wonder how the MPT handles non-existing accounts VS existing accounts with default values in their fields.

Reply:
Yes, the accounts are deleted when root is [recomputed](https://github.com/ethereum/go-ethereum/blob/master/core/state/statedb.go#L861).

Non-existing accounts are not stored in the trie (no leaf at the position specified by the account address), while for the account with default values there is a leaf which contains default values. However, in MPT circuit, when there is no account, a placeholder leaf with default values is set, so the lookup should contain default value as previous value. Because currently MPT circuit allows only one change at a time, there would be four lookups needed for the account creation. Just working on that actually. Let me prepare a hackmd note on that.

Reply:
I am not very familiar with the current mpt circuit. But just a remainder: it is very dangerous to allow two tries to have same semantic. I mean "However, in MPT circuit, when there is no account, a placeholder leaf with default values is set, so the lookup should contain default value as previous value.", maybe we should be very careful about the soundness if we allow both "non exist account(not exist in tree)" and "account with default value", where same semantic trees can have different roots.

Reply:
I think this EIP can be very useful when reasoning about empty accounts and how they are represented in the MPT: https://eips.ethereum.org/EIPS/eip-161

Reply:
> I am not very familiar with the current mpt circuit. But just a remainder: "However, in MPT circuit, when there is no account, a placeholder leaf with default values is set, so the lookup should contain default value as previous value.", maybe we should be very careful about the soundness if we allow both "non exist account(not exist in tree)" and "account with default value", where same semantic trees can have different roots.

Agree, but I think I just didn't phrase it clearly. The two tries (with non-existent account and account with default values) are not equivalent in the MPT circuit, it's just that in case of a non-existent account a placeholder leaf is added not to break the MPT circuit layout. This placeholder leaf is then ignored in the circuit, but the circuit of course checks that this leaf is actually empty - this check happens in the branch (below this branch a placeholder leaf resides).